lib/krb5/pac.c

```diff
@@ -1201,3 +1306,223 @@ _krb5_pac_sign(krb5_context context,
     return ret;
 }
+
+KRB5_LIB_FUNCTION krb5_error_code KRB5_LIB_CALL
+_krb5_pac_get_kdc_checksum_info(krb5_context context,
+				krb5_pac pac,
+				krb5_cksumtype *cstype,
+				uint16_t *rodc_id)
+{
+    krb5_error_code ret;
+    krb5_storage *sp = NULL;
+    const struct PAC_INFO_BUFFER *sig;
+    size_t cksumsize, prefix;
+    uint32_t type = 0;
+
+    *cstype = 0;
+    *rodc_id = 0;
+
+    sig = pac->privsvr_checksum;
+    if (sig == NULL) {
+	krb5_set_error_message(context, EINVAL, "PAC missing kdc checksum");
+	return EINVAL;
+    }
+
+    sp = krb5_storage_from_mem((char *)pac->data.data + sig->offset_lo,
+			       sig->buffersize);
+    if (sp == NULL)
+	return krb5_enomem(context);
+
+    krb5_storage_set_flags(sp, KRB5_STORAGE_BYTEORDER_LE);
+
+    ret = krb5_ret_uint32(sp, &type);
+    if (ret) {
+	krb5_storage_free(sp);
+	return ret;
+    }
+
+    ret = krb5_checksumsize(context, type, &cksumsize);
+    if (ret) {
+	krb5_storage_free(sp);
+	return ret;
+    }
+
+    *cstype = type;
+
+    prefix = krb5_storage_seek(sp, 0, SEEK_CUR);
+
+    if (cksumsize == (sig->buffersize - prefix)) {
+	krb5_storage_free(sp);
+	return 0;
+    }
+
+    if (cksumsize != (sig->buffersize - prefix - 2)) {
+	krb5_storage_free(sp);
+	return EINVAL;
+    }
+
+    krb5_storage_seek(sp, cksumsize, SEEK_CUR);
+
+    ret = krb5_ret_uint16(sp, rodc_id);
+
+    krb5_storage_free(sp);
+
+    return ret;
+}
+
+
+KRB5_LIB_FUNCTION krb5_error_code KRB5_LIB_CALL
+_krb5_kdc_pac_ticket_parse(krb5_context context,
+			   EncTicketPart *tkt,
+			   krb5_boolean *signedticket,
+			   krb5_pac *ppac)
+{
+    AuthorizationData *ad = tkt->authorization_data;
+    unsigned i, j;
+    krb5_pac pac = NULL;
+    size_t len = 0;
+    krb5_error_code ret;
+
+    *signedticket = FALSE;
+    *ppac = NULL;
+
+    if (ad == NULL || ad->len == 0)
+	return 0;
+
+    for (i = 0; i < ad->len; i++) {
+	AuthorizationData child;
+
+	if (ad->val[i].ad_type != KRB5_AUTHDATA_IF_RELEVANT)
+	    continue;
+
+	ret = decode_AuthorizationData(ad->val[i].ad_data.data,
+				       ad->val[i].ad_data.length,
+				       &child,
+				       NULL);
+	if (ret) {
+	    krb5_set_error_message(context, ret, "Failed to decode "
+				   "IF_RELEVANT with %d", ret);
+	    return ret;
+	}
+
+	for (j = 0; j < child.len; j++) {
+
+	    if (child.val[j].ad_type == KRB5_AUTHDATA_WIN2K_PAC) {
+		static unsigned char single_zero = '\0';
+		krb5_data adifr_data = ad->val[i].ad_data;
+		krb5_data pac_data = child.val[j].ad_data;
+		krb5_data recoded_adifr;
+
+		/* Found PAC */
+		ret = krb5_pac_parse(context,
+				     pac_data.data,
+				     pac_data.length,
+				     &pac);
+		if (ret) {
+		    free_AuthorizationData(&child);
+		    return ret;
+		}
+
+		if (pac->ticket_checksum == NULL) {
+		    free_AuthorizationData(&child);
+
+		    *ppac = pac;
+		    return 0;
+		}
+
+		child.val[j].ad_data.length = 1;
+		child.val[j].ad_data.data = &single_zero;
+
+		ASN1_MALLOC_ENCODE(AuthorizationData, recoded_adifr.data,
```

Comment:
Maybe a little comment saying, re-encoded the ticket with the PAC replaced with a single NUL byte, to be used as input to the ticket signature.